src/net/dns.cc

```diff
@@ -837,16 +839,34 @@ class net::dns_resolver::impl
                     f = e.tcp.out->write(std::move(p));
                     break;
                 case type::udp:
-                    f = e.udp.channel.send(e.udp.dst, std::move(p));
-                    break;
+                    // always chain UDP sends
+                    e.udp.f = e.udp.f.then([&e, p = std::move(p)]() mutable {
+                        auto f = e.udp.channel.send(e.udp.dst, std::move(p));;
+                        return f;
+                    }).finally([fd, me = shared_from_this()] {
+                        me->release(fd);
+                    });
+                    // if we have a fast-fail, give error.
+                    if (e.udp.f.failed()) {
+                        try {
+                            e.udp.f.get();
+                        } catch (std::system_error& e) {
+                            errno = e.code().value();
+                        } catch (...) {
+                        }
+                        return -1;
```

Comment:
For cleanliness sake, reinitialize f here. I don't think it will break but futures are use-once things (future::get() with a T will move the T out).

Reply:
Right